test/module/shared_model/validators/field_validator_test.cpp

```diff
@@ -0,0 +1,747 @@
+          false,
+          "Wrongly formed domain_id, passed value: '+math+'",
+      },
+      {
+          "id_too_long",
+          [&] { domain_id = "somelongname"; },
+          false,
+          "Wrongly formed domain_id, passed value: 'somelongname'",
+      },
+  };
+
+  std::vector<FieldTestCase> asset_name_test_cases{
+      {
+          "valid_name",
+          [&] { asset_name = "admin"; },
+          true,
+          "",
+      },
+      {
+          "empty_string",
+          [&] { asset_name = ""; },
+          false,
+          "Wrongly formed asset_name, passed value: ''",
+      },
+      {
+          "illegal_char",
+          [&] { asset_name = "+math+"; },
+          false,
+          "Wrongly formed asset_name, passed value: '+math+'",
+      },
+      {
+          "name_too_long",
+          [&] { asset_name = "somelongname"; },
+          false,
+          "Wrongly formed asset_name, passed value: 'somelongname'",
+      },
+  };
+
+  std::vector<FieldTestCase> permissions_test_cases{
+      {
+          "valid_role",
+          [&] {
+            role_permission = iroha::protocol::RolePermission::can_append_role;
+          },
+          true,
+          "",
+      },
+  };
+
+  std::vector<FieldTestCase> tx_counter_test_cases{
+      {
+          "valid_counter",
+          [&] { counter = 5; },
+          true,
+          "",
+      },
+      {
+          "zero_counter",
+          [&] { counter = 0; },
+          false,
+          "Counter should be > 0",
+      },
+  };
+  std::vector<FieldTestCase> created_time_test_cases{
+      {
+          "valid_time",
+          [&] { created_time = iroha::time::now(); },
+          true,
+          "",
+      },
+  };
+
+  std::vector<FieldTestCase> detail_test_cases{
+      {
+          "valid_detail_key",
+          [&] { detail_key = "happy"; },
+          true,
+          "",
+      },
+      {
+          "empty_string",
+          [&] { detail_key = ""; },
+          false,
+          "Wrongly formed key, passed value: ''",
+      },
+      {
+          "illegal_char",
+          [&] { detail_key = "hi*there"; },
+          false,
+          "Wrongly formed key, passed value: 'hi*there'",
+      },
+  };
+  std::vector<FieldTestCase> &key_test_cases = detail_test_cases;
+
+  // no constraints yet
+  std::vector<FieldTestCase> precision_test_cases;
+  std::vector<FieldTestCase> permission_test_cases;
+  std::vector<FieldTestCase> value_test_cases;
+  std::vector<FieldTestCase> quorum_test_cases;
+  std::vector<FieldTestCase> description_test_cases;
+  std::vector<FieldTestCase> signature_test_cases;
+  std::vector<FieldTestCase> tx_hashes_test_cases;
+
+  /**************************************************************************/
+
+  // register validation function and test cases
+  std::unordered_map<std::string, FieldTest> field_validators{
+      // Command fields
+      {"account_id",
+       {[&] {
+          validation::ReasonsGroupType reason;
```

Comment:
Also looks like a shortcut is required, so that `reason` lines are not duplicated.

Reply:
I was thinking about passing a reference to the method to function, but non-static class members cannot be passed that way, I can wrap them in lambda, but the amount of code would not decrease in this case.